62.SchusslerTest/shader.frag

```diff
@@ -91,37 +89,56 @@ vec3 evalBRDF(vec3 wi, vec3 wo, vec3 N) {
 void main()
 {
     const float Intensity = 20.0;
+    const float THRESHOLD = 1.0 - 1e-5;
 
+    vec3 dFdxPos = dFdx(Pos);
+    vec3 dFdyPos = dFdy(Pos);
+    vec3 GN = normalize(cross(dFdxPos,dFdyPos));
+    vec3 N = normalize(Normal);
     vec3 L = LightPos-Pos;
-    vec3 Lnorm = normalize(L);
-    const float THRESHOLD = AlphaY+0.5;
-    float a2 = Alpha;
-    float a2formeta = AlphaY;
-    
+
     vec3 wi = normalize(CamPos - Pos);
-    vec3 wo = Lnorm;
-    const float deviation = dot(Normal,GNormal);
-    vec3 wt = -normalize(Normal-deviation*GNormal);
-    vec3 wdasho = normalize(reflect(-wo, wt));
-    vec3 wdashi = normalize(reflect(-wi, wt));
+    vec3 wo = normalize(L);
+    const float deviation = dot(N,GN);
+    if (deviation < 0) {
+        outColor = vec4(1.0, 0.0, 0.0, 1.0);
+        return;
+    } 
+    vec3 wt = -normalize(N-deviation*GN);
+    vec3 wdasho = reflect(wo, wt);
+    vec3 wdashi = reflect(wi, wt);
 
-    const vec3 albedo = vec3(0.5);
-    vec3 brdf_cos_wi_wo = evalBRDF(wi,wo,Normal);
-    vec3 brdf_cos_wi_wdasho = evalBRDF(wi,wdasho,Normal);
-    vec3 brdf_cos_wdashi_wo = evalBRDF(wdashi,wo,Normal);
-    vec3 brdf;
-    if (deviation<THRESHOLD) 
+    float a2 = Alpha;
+    const float NdotV = dot(GN,wi);
+    const float NdotL = dot(GN,wo);
+    if (NdotV<nbl_glsl_FLT_MIN || NdotL<nbl_glsl_FLT_MIN) {
+        outColor = vec4(0.0,0.0,0.0,1.0);
+        return;
+    }
+
+    vec3 brdf_cos_wi_wo = evalBRDF(wi,wo,N);
+    vec3 brdf_cos_wi_wdasho = evalBRDF(wi,wdasho,N);
+    vec3 brdf_cos_wdashi_wo = evalBRDF(wdashi,wo,N);
+    vec3 brdf = vec3(0.0);
+
+    if (deviation < THRESHOLD) 
     {
-        brdf = lambdap(wi, wt, Normal, GNormal) * (
-            G1(wo, Normal, a2formeta) * brdf_cos_wi_wo +
-            G1(wo, wt, a2formeta) * (1.0 - G1(wdasho, Normal, a2formeta)) * brdf_cos_wi_wdasho
-        ) + lambdat(wi, wt, Normal, GNormal) * brdf_cos_wdashi_wo * G1(wo, Normal, a2formeta);
+        if (dot(wi, GN) >= 0.0) brdf += lambdap(wi, wt, N, GN) * G1(wo, N, wt, N, GN) * brdf_cos_wi_wo;
+        if (dot(wi, GN) >= 0.0) brdf += lambdap(wi, wt, N, GN) * G1(wo, wt,  wt, N, GN) 
+            * (1.0 - G1(wdasho, N,  wt, N, GN)) * brdf_cos_wi_wdasho;
+        if (dot(wdashi, GN) >= 0.0) brdf += lambdat(wi, wt, N, GN) * brdf_cos_wdashi_wo * G1(wo, N,  wt, N, GN);
```

Comment:
the `if` statements are wrong, checks should be inside the `evalBRDF` functions (for now) and you need to check both `wi` and `wo` against shading normal, not geometrical normal!